### ISSUE TITLE
Nondeterministic game tree during MCTS expansion

### ISSUE THREAD

**Issue by melvinzhang** [magarena/magarena#307]
```
1019. UntapAction (Badlands) = -5
1020. UntapAction (Mountain) = -5
1021. UntapAction (Mountain) = -5
1022. UntapAction (Ravenous Skirge) = -10
1023. UntapAction (Swamp) = -5
1024. DrawAction = -335
1025. AddEventAction (EVENT: NONE  ...) = 0
1026. ExecuteFirstEventAction = 0
1027. AddEventAction (EVENT: NONE  Play a card or ability.) = 0
1028. ExecuteFirstEventAction = 0
1029. AddEventAction (EVENT: Shuriken Pay {2}$. Choose how to pay the mana cost.) = 0
1030. AddEventAction (EVENT: Shuriken Play activated ability of Shuriken. ) = 0
1031. ExecuteFirstEventAction = 0
1032. TapAction (Badlands) = 5
1033. TapAction (Mountain) = 5
1034. ExecuteFirstEventAction = 0
1035. PutItemOnStackAction (MagicAbilityOnStack,Shuriken) = 1
1036. AddEventAction (EVENT: Shuriken  Choose target creature you control.) = 0
1037. ExecuteFirstEventAction = 0
1038. AddEventAction (EVENT: NONE  Play a card or ability.) = 0
1039. ExecuteFirstEventAction = 0
1040. AddEventAction (EVENT: NONE  Play a card or ability.) = 0
1041. ExecuteFirstEventAction = 0
1042. StackResolveAction = 0
1043. PlayAbilityAction = 0
1044. AttachAction = 0
1045. AddEventAction (EVENT: NONE  Play a card or ability.) = 0
1046. ExecuteFirstEventAction = 0
1047. AddEventAction (EVENT: NONE  Play a card or ability.) = 0
1048. ExecuteFirstEventAction = 0
1049. AddEventAction (EVENT: NONE  Declare attackers.) = 0
Score = 1541

Exception from controller.runGame: Index: 2, Size: 2
java.lang.IndexOutOfBoundsException: Index: 2, Size: 2
        at java.util.ArrayList.rangeCheck(ArrayList.java:635)
        at java.util.ArrayList.get(ArrayList.java:411)
        at magic.ai.MCTSAI.growTree(MCTSAI.java:410)
        at magic.ai.MCTSAI.TreeUpdate(MCTSAI.java:245)
        at magic.ai.MCTSAI$1.run(MCTSAI.java:151)
```

**Comment by melvinzhang** [comment]
From http://www.slightlymagic.net/forum/viewtopic.php?f=82&t=17255&p=180981#p180978
```
 1234. CombatDamageAction = 0                                                                                                                                                                  
 1235. AddEventAction (EVENT: NONE  Play a card or ability.) = 0                                                                                                                               
 1236. ExecuteFirstEventAction = 0                                                                                                                                                             
 1237. AddEventAction (EVENT: NONE  Play a card or ability.) = 0                                                                                                                               
 1238. ExecuteFirstEventAction = 0                                                                                                                                                             
 1239. CombatDamageAction = 0                                                                                                                                                                  
 1240. DealDamageAction = 0                                                                                                                                                                    
 1241. AddEventAction (EVENT: NONE  Play a card or ability.) = 0                                                                                                                               
 1242. ExecuteFirstEventAction = 0                                                                                                                                                             
 1243. AddEventAction (EVENT: NONE  Play a card or ability.) = 0                                                                                                                               
 1244. ExecuteFirstEventAction = 0                                                                                                                                                             
 1245. RemoveFromCombatAction (Permanent #324) = 0                                                                                                                                             
 1246. ChangeStateAction = 0                                                                                                                                                                   
 1247. ChangeStateAction = 0                                                                                                                                                                   
 1248. ChangeStateAction = 0                                                                                                                                                                   
 1249. AddEventAction (EVENT: NONE  Play a card or ability.) = 0                                                                                                                               
 Score = 4579                                                                                                                                                                                  
                                                                                                                                                                                               
 Exception from controller.runGame: Index: 4, Size: 2                                                                                                                                          
 java.lang.IndexOutOfBoundsException: Index: 4, Size: 2                                                                                                                                        
         at java.util.ArrayList.rangeCheck(ArrayList.java:635)                                                                                                                                 
         at java.util.ArrayList.get(ArrayList.java:411)                                                                                                                                        
         at magic.ai.MCTSAI.growTree(MCTSAI.java:410)                                                                                                                                          
         at magic.ai.MCTSAI.TreeUpdate(MCTSAI.java:245)                                                                                                                                        
         at magic.ai.MCTSAI$1.run(MCTSAI.java:151)                                                                                                                                             
         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)                                                                                                            
         at java.util.concurrent.FutureTask.run(FutureTask.java:262)                                                                                                                           
         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)                                                                                                    
         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)                                                                                                    
         at java.lang.Thread.run(Thread.java:744)          
```

**Comment by ShawnieBoy** [comment]
At least we now have a replicable test for this crash :grinning:

**Comment by melvinzhang** [comment]
One possible cause of non deterministic game tree is dynamically created MagicActivation. MagicActivation is sorted by hashcode, if a MagicActivation is created fresh during the game (such as with the previous implementation of Shuriken) it will have a different hashcode each time and be sorted differently in the list of choices.

**Comment by melvinzhang** [comment]
The second crash log that mentions "Permanent #324" is most likely due to manifest creating a MagicActivation dynamically. This was fixed in  2fc4dfc70b3f
